components/learn/modules/ROOT/pages/on-gsn.adoc

```diff
@@ -1,9 +1,338 @@
+  );
+}
+
+export default App;
+
+----
+
+
+NOTE: You can pass a `dev: true` flag to the `gsn` options when setting up the provider. This will use the xref:gsn-provider::index.adoc[GSNDevProvider] instead of the regular GSN provider. This is a provider set up specifically for testing or development, and it _does not require a relayer to be running_ to work. This can make development easier, but it will feel less like the actual GSN experience. If you want to use an actual relayer, you can run `npx oz-gsn run-relayer` locally (see the xref:gsn-helpers::preparing-a-testing-environment.adoc[OpenZeppelin GSN helpers] for more info).
+
+Great! We can now fire up our application running `npm start` from within the `client` folder. Remember to keep both your ganache and relayer up and running. You should be able to send transactions to your `Counter` contract without having to use MetaMask or have any ETH at all!
+
+[[moving-to-testnet]]
+=== Moving to a testnet
+
+It is not too impressive sending a local transaction in your ganache network, where you already have a bunch of fully-funded accounts. To witness the GSN at its full potential, let's move our application to the Rinkeby testnet. If you later want to go onto mainnet, the instructions are the same.
+
+Let's start by deploying our `Counter` contract to Rinkeby. You will need an account with some Rinkeby Ether for this, which you will have to register in your `network.js` file. Take a look at xref:public-deploy.adoc[deploying to a public network] guide for more information.
+
+[source,console]
+----
+$ openzeppelin create
+✓ Compiled contracts with solc 0.5.9 (commit.e560f70d)
+? Pick a contract to instantiate: Counter
+? Pick a network: rinkeby
+✓ Added contract Counter
+✓ Contract Counter deployed
+? Call a function to initialize the instance after creating it?: Yes
+? Select which function * initialize()
+✓ Setting everything up to create contract instances
+✓ Instance created at 0xCfEB869F69431e42cdB54A4F4f105C19C080A601
+----
+
+
+The next step will be to instruct our (d)app to connect to a Rinkeby node instead of to the local network. Change the `PROVIDER_URL` in your `App.js` to do this using, for instance, an Infura Rinkeby endpoint. At this point, you will also want to pass in a config object as we will be using a real GSN provider rather than our developer environment, and our config options give us more control over things such as the gas price we are willing to pay. For production (d)apps, you will want to configure this to your requirements. 
+
+[source,javascript]
+----
+
+import { useWeb3Network, useEphemeralKey } from "@openzeppelin/network/react";
+
+// inside App.js#App()
+const context = useWeb3Network('https://rinkeby.infura.io/v3/' + INFURA_API_TOKEN, {
+  gsn: { signKey: useEphemeralKey() }
+});
+----
+ 
+Redeploy using the cli command `oz create`, selecting `Rinkeby` as the network, and copy the address returned at the end (You will need it later to fund your contract!).
```

Comment:
```suggestion
```

Reply:
Why removed this bit?

Reply:
We've deployed two paragraphs ago, this instruction was repeated (this was less obvious before the suggestions were applied)

Reply:
Oops, good point!